### PR TITLE
[SPARK-29657][CORE] Iterator spill supporting radix sort with null prefix

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
Support ```ChainedIterator``` with ```SortedIterator``` spill

### Why are the changes needed?
In the case of radix sort, when the ```insertRecord``` part of the keyPrefix is null, the iterator type returned by ```getSortedIterator``` is ```ChainedIterator```.
Currently ```ChainedIterator``` does not support spill, causing ```UnsafeExternalSorter``` to take up a lot of execution memory, ```allocatePage``` fails, throw ```SparkOutOfMemoryError``` Unable to acquire xxx bytes of memory, got 0

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
add ut

### CODE REVIEW

**Comment by cxzl25** [comment]
```UnsafeInMemorySorter#getSortedIterator```  
Use radix sort, some keyPrefix has null, return ChainedIterator

https://github.com/apache/spark/blob/44a27bdccdc39d5394ee95d935455eb7ff4b84c2/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeInMemorySorter.java#L364-L376

```readingIterator.spill```
https://github.com/apache/spark/blob/44a27bdccdc39d5394ee95d935455eb7ff4b84c2/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java#L519-L524


The following is a log of an error we encountered in the production environment.

[Executor task launch worker for task 66055] INFO TaskMemoryManager: Memory used in task 66055
[Executor task launch worker for task 66055] INFO TaskMemoryManager: Acquired by org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter@39dd866e: 64.0 KB
[Executor task launch worker for task 66055] INFO TaskMemoryManager: Acquired by org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter@74d17927: 4.6 GB
[Executor task launch worker for task 66055] INFO TaskMemoryManager: Acquired by org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter@31478f9c: 61.0 MB
[Executor task launch worker for task 66055] INFO TaskMemoryManager: 0 bytes of memory were used by task 66055 but are not associated with specific consumers
[Executor task launch worker for task 66055] INFO TaskMemoryManager: 4962998749 bytes of memory are used for execution and 2218326 bytes of memory are used for storage
[Executor task launch worker for task 66055] ERROR Executor: Exception in task 42.3 in stage 29.0 (TID 66055)
SparkOutOfMemoryError: Unable to acquire 3436 bytes of memory, got 0

**Comment by cxzl25** [comment]
@JoshRosen @davies 
Can you look at this pr when you are free?  Thank you.

**Comment by cxzl25** [comment]
@cloud-fan Can you look at this question?
Can really improve the success rate in some extreme scenarios.

**Comment by cxzl25** [comment]
@wangyum @HyukjinKwon @srowen 
Can you look at this pr? Thank you.
This pr was raised for a while but was not confirmed.

**Comment by srowen** [comment]
I don't know enough to evaluate the change. It looks like it touches some important code.

**Comment by cloud-fan** [comment]
OK to test

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 1]
```diff
@@ -518,19 +520,52 @@ public int getNumRecords() {
 
     public long spill() throws IOException {
```
Comment (cloud-fan) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: low-level code is hard to review as nobody remembers all the details all the time. Can you briefly explain the spill process to help people review?

Comment (cxzl25) [reply core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: Before [SPARK-14851](https://issues.apache.org/jira/browse/SPARK-14851), ```UnsafeInMemorySorter.getSortedIterator()```, the returned type is ```UnsafeInMemorySorter.SortedIterator```.

Now, when the radio sort is turned on and there are some empty values, the return type is ```UnsafeExternalSorter.ChainedIterator(UnsafeInMemorySorter.SortedIterator)```.
In ```UnsafeExternalSorter.SpillableIterator#spill```, this situation cannot be spilled.


When ```UnsafeExternalSorter#getSortedIterator``` is called, part of the last written data is saved in memory.

When the task needs more execution memory, the spill fails.

TaskMemoryManager#acquireExecutionMemory->UnsafeExternalSorter#spill->UnsafeExternalSorter.SpillableIterator#spill

**Pull request closed without merging**